x-pack/plugins/lists/server/services/exception_lists/exception_list_client.mock.ts

```diff
@@ -47,11 +63,189 @@ export class ExceptionListClientMock extends ExceptionListClient {
+
+  return {
+    comments,
+    description,
+    entries,
+    itemId,
+    listId,
+    meta,
+    name,
+    namespaceType,
+    osTypes,
+    tags,
+    type,
+  };
+};
+
+export const getUpdateExceptionListItemOptionsMock = (): UpdateExceptionListItemOptions => {
+  const { comments, entries, itemId, namespaceType, name, osTypes, description, meta, tags, type } =
+    getCreateExceptionListItemOptionsMock();
+
+  return {
+    _version: undefined,
+    comments,
+    description,
+    entries,
+    id: ID,
+    itemId,
+    meta,
+    name,
+    namespaceType,
+    osTypes,
+    tags,
+    type,
+  };
+};
+
+export const getExceptionListSoSchemaMock = (
+  overrides: Partial<ExceptionListSoSchema> = {}
+): ExceptionListSoSchema => {
+  /* eslint-disable @typescript-eslint/naming-convention */
+  const {
+    comments,
+    created_at,
+    created_by,
+    description,
+    entries,
+    item_id,
+    list_id,
+    meta,
+    name,
+    os_types,
+    tags,
+    tie_breaker_id,
+    type,
+    updated_by,
+  } = getExceptionListItemSchemaMock();
+  /* eslint-enable @typescript-eslint/naming-convention */
+
+  const soSchema: ExceptionListSoSchema = {
+    comments,
+    created_at,
+    created_by,
+    description,
+    entries,
+    immutable: undefined,
+    item_id,
+    list_id,
+    list_type: 'item',
+    meta,
+    name,
+    os_types,
+    tags,
+    tie_breaker_id,
+    type,
+    updated_by,
+    version: undefined,
+    ...overrides,
+  };
+
+  return soSchema;
+};
+
+/**
+ * Returns a Saved Object with the `ExceptionListSoSchema` as the attributes
+ * @param attributesOverrides
+ * @param savedObjectOverrides
+ */
+export const getExceptionListItemSavedObject = (
+  attributesOverrides: Partial<ExceptionListSoSchema> = {},
+  savedObjectOverrides: Partial<Omit<SavedObject, 'attributes'>> = {}
+): SavedObject<ExceptionListSoSchema> => {
+  return {
+    attributes: getExceptionListSoSchemaMock(attributesOverrides),
+    coreMigrationVersion: undefined,
+    error: undefined,
+    id: ID,
+    migrationVersion: undefined,
+    namespaces: undefined,
+    originId: undefined,
+    references: [],
+    type: getSavedObjectType({ namespaceType: 'agnostic' }),
+    updated_at: DATE_NOW,
+    version: _VERSION,
+    ...savedObjectOverrides,
+  };
+};
+
+/**
+ * Returns a saved objects client mock that includes method mocks to handle working with the exceptions list client.
+ * @param [soClient] can be provided on input and its methods will be mocked for exceptions list only and will preserve existing ones for other types
+ */
+export const getExceptionListSavedObjectClientMock = (
+  soClient: ReturnType<typeof savedObjectsClientMock.create> = savedObjectsClientMock.create()
+): ReturnType<typeof savedObjectsClientMock.create> => {
+  // mock `.create()`
+  const origCreateMock = soClient.create.getMockImplementation();
+  // @ts-expect-error
+  soClient.create.mockImplementation(async (...args) => {
```

Comment:
`ts-expect-error` is the last possible thing developers should be doing as in I can do something as bad as misname something and the code will take it. It literally turns off all of the type checks for a line and is considered as only a last resort or if someone is doing an incredibly large "sweep" of code bases where they don't understand the logic well enough to make a decision.

If someone renames `soClient.create` for example, they have to figure out where it broke by looking at the failed unit tests (if they still fail) rather than getting a static analysis from TypeScript alerting us that it needs to be fixed.

Each time we add a `ts-expect-error`, it:
* counts as tech debt and should be avoidable in pretty much every case for new code including tests. 
* It's a ticking time bomb or at least less safe in that soon enough method names will change and a maintainer/developer will have to look at individual failed unit tests to sleuth out where something is going wrong rather than types telling them.
 
For example, we can do this here:

```ts
  // mock `.create()`
  const origCreateMock = soClient.create.getMockImplementation();
  if (origCreateMock == null) {
    throw new TypeError('origCreateMock should never be null or undefined during tests');
  }
```

And that will allow us to remove the `ts-expect-error`. If the method name changes, the type check will fail and if this does return an `undefined` we should see a clear stack trace and error message about what is going wrong.

Reply:
I don't know that this work-around will actually work. the problem is not the `origCreateMock` being null/undefined - that is actually a valid condition when there is no mock implementation (ex. `create: jest.fn()`).  the issue was the the function provided to `mockImplementation()`. I will try again to see if I can get the types right, but may end up casting it and removing the `ts-expect-error` is thats cool(?).

![image](https://user-images.githubusercontent.com/56442535/148113008-389182c2-f2f2-4289-b71c-2678aa894e18.png)

Reply:
Ok. I found that this works, so I hop the cast is ok (which essentially matches what the mock is set to initially anyway)

```typescript
  // mock `.create()`
  const origCreateMock = soClient.create.getMockImplementation();
  soClient.create.mockImplementation(async (...args) => {
    const [type, attributes] = args;

    if (isExceptionsListSavedObjectType(type)) {
      return getExceptionListItemSavedObject(attributes as ExceptionListSoSchema, { type });
    }

    if (origCreateMock) {
      return origCreateMock(...args);
    }

    return undefined as unknown as SavedObject;
  });
```